tools/val/val.cpp

```diff
@@ -33,73 +34,122 @@ or if the filename is "-", then the binary is read from standard input.
     const char* cur_arg = argv[argi];
     if ('-' == cur_arg[0]) {
-      if (0 == strcmp(cur_arg, "--version")) {
+      if (0 == strncmp(cur_arg, "--max-", 6)) {
+        if (argi + 1 < argc) {
+          spv_validator_limit limit_type;
+          if (spvParseUniversalLimitsOptions(cur_arg, &limit_type)) {
+            uint32_t limit = 0;
+            if (sscanf(argv[++argi], "%d", &limit)) {
+              spvValidatorOptionsSetUniversalLimit(options, limit_type, limit);
+            } else {
+              fprintf(stderr, "error: missing argument to %s\n", cur_arg);
+              continue_processing = false;
+              return_code = 1;
+            }
+          } else {
+            fprintf(stderr, "error: unrecognized option: %s\n", cur_arg);
+            continue_processing = false;
+            return_code = 1;
+          }
+        } else {
+          fprintf(stderr, "error: Missing argument to %s\n", cur_arg);
+          continue_processing = false;
+          return_code = 1;
+        }
+      } else if (0 == strcmp(cur_arg, "--version")) {
         printf("%s\n", spvSoftwareVersionDetailsString());
         printf("Targets:\n  %s\n  %s\n",
                spvTargetEnvDescription(SPV_ENV_UNIVERSAL_1_1),
                spvTargetEnvDescription(SPV_ENV_VULKAN_1_0));
-        return 0;
+        continue_processing = false;
+        return_code = 0;
       } else if (0 == strcmp(cur_arg, "--help") || 0 == strcmp(cur_arg, "-h")) {
         print_usage(argv[0]);
-        return 0;
+        continue_processing = false;
+        return_code = 0;
       } else if (0 == strcmp(cur_arg, "--target-env")) {
         if (argi + 1 < argc) {
           const auto env_str = argv[++argi];
           if (!spvParseTargetEnv(env_str, &target_env)) {
             fprintf(stderr, "error: Unrecognized target env: %s\n", env_str);
-            return 1;
+            continue_processing = false;
+            return_code = 1;
           }
         } else {
           fprintf(stderr, "error: Missing argument to --target-env\n");
-          return 1;
+          continue_processing = false;
+          return_code = 1;
         }
       } else if (0 == cur_arg[1]) {
         // Setting a filename of "-" to indicate stdin.
         if (!inFile) {
           inFile = cur_arg;
         } else {
           fprintf(stderr, "error: More than one input file specified\n");
-          return 1;
+          continue_processing = false;
+          return_code = 1;
         }
-
       } else {
         print_usage(argv[0]);
-        return 1;
+        continue_processing = false;
+        return_code = 1;
       }
     } else {
       if (!inFile) {
         inFile = cur_arg;
       } else {
         fprintf(stderr, "error: More than one input file specified\n");
-        return 1;
+        continue_processing = false;
+        return_code = 1;
       }
     }
   }
 
+  // Exit if command line parsing was not successful.
+  if (!continue_processing) {
```

Comment:
This is fine.   However, another two ways to handle this:
1. Don't care about leaking the object because this is the main function and it's going to exit soon.  :-)  Slightly ugly, but if you do that then add a comment saying "Deliberately leaking this object."
2. Do options object management unconditionally in this top level main function, but then wrap the "actually do stuff" remainder of the code in a function.  This is more complex than what you have.